Support zod@4

Bram-dc opened the pull request:
This PR is far from ready to be merged, mostly due to bugs in the early release of zod v4. Also the way zod handles issues has changed so we might also want to do it differently in this package.

This PR has big breaking changes since it uses a different zod-to-json (built into zod@4)

Tests are failing, due to bugs in zod

Bram-dc commented on test/__snapshots__/fastify-swagger.spec.ts.snap:
```diff
@@ -128,6 +116,7 @@ exports[`transformer > should generate ref correctly 1`] = `
   "components": {
     "schemas": {
       "Token": {
+        "id": "Token",
```
This is not valid in the OpenAPI spec. The best solution is to fix this upstream. Removing it manually feels dirty.

Bram-dc replied:
This has been resolved by not using the global registry, more testing has to be done on the stable release to ensure this also wont happen using the global registry.

Bram-dc commented on test/__snapshots__/fastify-swagger.spec.ts.snap:
```diff
@@ -41,15 +33,13 @@ exports[`transformer > generates types for fastify-swagger correctly 1`] = `
                     "type": "string",
                   },
                   "seed": {
-                    "exclusiveMinimum": true,
```
z.int().positive() is currently undefined, so this is a workaround.

See: https://github.com/colinhacks/zod/issues/4087

kibertoad replied:
this was fixed

Bram-dc commented on src/errors.ts:
```diff
@@ -60,9 +60,9 @@ export const hasZodFastifySchemaValidationErrors = (
 
 export const createValidationError = (error: ZodError): ZodFastifySchemaValidationError[] =>
-  error.errors.map((issue) => ({
+  error.issues.map((issue) => ({
```
Issues have changed, we should not access error.issues here.

See: https://v4.zod.dev/v4/changelog#deprecates-addissue-and-addissues

Bram-dc commented on src/core.ts:
```diff
@@ -93,20 +100,29 @@ export const jsonSchemaTransform = createJsonSchemaTransform({
-  ({ schemas }: { schemas: Record<string, z.ZodTypeAny> }) =>
-  (
-    input:
-      | { swaggerObject: Partial<OpenAPIV2.Document> }
```
The type `OpenAPIV2.Document` mismatched with `fastity-swagger`

Bram-dc replied:
Fixed in: https://github.com/fastify/fastify-swagger/pull/866

Bram-dc commented:
@kibertoad @turkerdev Could we have a nightly release or beta release so I can do some integrated testing? "npm i https://github.com/turkerdev/fastify-type-provider-zod" does not work with this current setup.

kibertoad commented:
@Bram-dc can you try using latest beta version? Wonder if it addressed some of the issues that you have encountered.

Bram-dc commented:
> @Bram-dc can you try using latest beta version? Wonder if it addressed some of the issues that you have encountered.

Yes it did resolve most of the issues, the only one causing our tests to fail is the "id" fields being added to the schema when using the global registry. There is also different behaviour when using `z.globalRegistry` and `z.registry()`. I might create a PR for that in the zod github.

kibertoad commented:
@Bram-dc Can you update this for the final 3.25.3 release? We could then merge it after it is green

adri1wald commented:
just popping by here to say i'm so excited and so thankful for this release! it'll let us finally fix the excruciatingly slow type checking in our repo 🤩

kibertoad commented:
@Bram-dc What are the outstanding zod bugs/issues that need to be addressed before we can get to green? Also, can I help with anything?

Bram-dc commented:
I think most issues are fixed, also zod 3.25 has some weird cjs imports, we need to fix. Me might need to use a different compiler in our project. But I think zod will fix their issues so I will wait for those to come. 

Also we should write a few extra zod-to-json function (we can extend behaviour). Mainly that the built into zod one is very minimal. Fastify tranforms Dates for example to iso strings. The built in zod-to-openapi returns an error as dates cannot be represented as a schema.

Bram-dc commented:
Waiting for this PR to be accepted: https://github.com/colinhacks/zod/pull/4525

Bram-dc commented:
> I think most issues are fixed, also zod 3.25 has some weird cjs imports we need to fix. We might need to use a different compiler in our project (maybe rollup?). But I think zod will fix their issues so I will wait for those to come.
> 
> Also we should write a few extra zod-to-json tranform functions (we can extend behaviour). Mainly that the built into zod one is very minimal. Fastify tranforms Dates for example to iso strings. The built in zod-to-openapi returns an error as dates cannot be represented as a schema.

Done and done

kibertoad commented:
@Bram-dc new version with the fix was released

gabriel-bianchessi commented:
Hey guys, great job migrating to zod v4. Do you have an approximate release date in mind? If I can contribute with something, let me know. :)

thelinuxlich commented:
thanks for the work @Bram-dc !

kibertoad commented:
I'll try to release on Sunday

kibertoad commented:
@Bram-dc Looks like the transition to dual-publishing wasn't entirely smooth, I'm getting the following error in an ESM project now:

```
Error: Cannot find module 'C:\sources\node\shared-ts-libs4\packages\app\fastify-api-contracts\node_modules\fastify-type-provider-zod\dist\esm\core' imported from C:\sources\node\shared-ts-libs4\packages\app\fastify-api-contracts\node_modules\fastify-type-provider-zod\dist\esm\index.js
```

Import looks like this:

```ts
import {
  type ZodTypeProvider,
  serializerCompiler,
  validatorCompiler,
} from 'fastify-type-provider-zod'
```

kibertoad commented:
@Bram-dc This is also very unhappy: https://arethetypeswrong.github.io/?p=fastify-type-provider-zod%405.0.0

kibertoad commented:
@Bram-dc Can you check https://github.com/turkerdev/fastify-type-provider-zod/pull/167?
